### ISSUE TITLE
`\ ` as multiline line continuation

### ISSUE THREAD

**Issue by isinyaaa** [hjson/hjson#106]
Many programming languages provide support to continue a line by using `\ ` at its end. E.g.:
```py
"this is a sentence \
that continues in the line below.
```
Which should be interpreted as `"this is a sentence that continues in the line below"`.

**Comment by trobro** [comment]
I don't think we will ever again make any changes to the Hjson syntax definition, it is fairly set in stone by now.

In Hjson triple quotes are used for multiline strings:

```
haiku:
  '''
  My half empty glass,
  I will fill your empty half.
  Now you are half full.
  '''
```

This evaluates to the string "My half empty glass,\nI will fill your empty half.\nNow you are half full."

**Comment by NuSkooler** [comment]
@isinyaaa  appears to be asking for a way to split a **single** line vs inserting line feeds. I'm not sure if this desired for HJSON but it's a different feature if I'm understanding correctly.

```
haiku:
'''
My half \
empty glass,
I will fill your empty half.
Now you are half full.
'''
```

**Comment by trobro** [comment]
True, poor reading by me.

**Comment by AlpyneDreams** [comment]
I think this is a very good idea because it'd mean Hjson would be syntactically fully compatible with [JSON5](https://json5.org/), which has this feature. The only other feature of JSON5 not supported by Hjson is extended numbers like infinity, NaN, hex numbers, etc., which Hjson already parses as strings, generally not breaking syntax.

**Comment by trobro** [comment]
@AlpyneDreams fair point, but existing Hjson files might already contain strings like this:

```
command:
  '''
  cmake \
    -Dhjson_DIR="C:/folderB/hjson-cpp/build64_VS15" \
    -DQt5_DIR="C:/folderA/qt/lib/cmake/Qt5" \
    -DCMAKE_GENERATOR_PLATFORM=x64 \
    ..
  '''
```

I would not want that string to be on a single line after a read-write cycle in Hjson. Also, there are currently no escape sequences at all in Hjson multiline strings, would be confusing to make an exception for backslash as line continuation.

**Comment by isinyaaa** [comment]
@trobro I guess a parse flag should be useful in this case, as it would allow compatibility as well as the use case you provided.

**Comment by AlpyneDreams** [comment]
> @AlpyneDreams fair point, but existing Hjson files might already contain strings like this:
> 
> ```
> command:
>   '''
>   cmake \
>     -Dhjson_DIR="C:/folderB/hjson-cpp/build64_VS15" \
>     -DQt5_DIR="C:/folderA/qt/lib/cmake/Qt5" \
>     -DCMAKE_GENERATOR_PLATFORM=x64 \
>     ..
>   '''
> ```
> 
> I would not want that string to be on a single line after a read-write cycle in Hjson. Also, there are currently no escape sequences at all in Hjson multiline strings, would be confusing to make an exception for backslash as line continuation.

Since JSON5 has no multiline triple-quote `'''` strings, the escaping of new lines could be applied only to double and single quote strings.  This would then be consistent with there being no escape sequences in Hjson multiline strings.

**Comment by AlpyneDreams** [comment]
> @AlpyneDreams fair point, but existing Hjson files might already contain strings like this:
> 
> ```
> command:
>   '''
>   cmake \
>     -Dhjson_DIR="C:/folderB/hjson-cpp/build64_VS15" \
>     -DQt5_DIR="C:/folderA/qt/lib/cmake/Qt5" \
>     -DCMAKE_GENERATOR_PLATFORM=x64 \
>     ..
>   '''
> ```
> 
> I would not want that string to be on a single line after a read-write cycle in Hjson. Also, there are currently no escape sequences at all in Hjson multiline strings, would be confusing to make an exception for backslash as line continuation.

Since JSON5 has no multiline triple-quote `'''` strings, the escaping of new lines could be applied only to double and single quote strings, not multi-line strings.  This would then be consistent with there being no escape sequences in Hjson multiline strings.

The syntax for them in regular strings would previously have been invalid Hjson so it shouldn't break any existing files.
```
    key: "Look, Mom! \
No \\n's!"
```



> @trobro I guess a parse flag should be useful in this case, as it would allow compatibility as well as the use case you provided.

Definitely a good idea, yeah.

**Comment by trobro** [comment]
I no longer have any objections, implementing this for single or double quoted strings only should be fine. Not for multiline and not for quoteless strings. I would support PR:s for this, but we need PR:s for all major supported languages before changing the syntax documentation. I can create one for hjson-cpp.

**Comment by trobro** [comment]
@isinyaaa would be great if you have time to implement for hjson-py! @dqsully what do you think of this?

**Comment by dqsully** [comment]
I think it's a good idea and should be easy to implement. And just to be clear, using `\` to wrap lines would only work in `""` and `''` strings, but not multiline or unquoted strings?

**Comment by trobro** [comment]
Great, I'll do a PR for hjson-cpp. Yes would only work in `""` and `''` strings, but not multiline or unquoted strings. That way there's no breaking change. Ending an unquoted string with a backslash will still produce a string that ends with a backslash.

**Comment by trobro** [comment]
...though now that I look at our syntax specification (https://hjson.github.io/syntax.html) I see that it does allow line feeds in `""` and `''` strings. I had not thought of that before, assumed it was not allowed. I'd better check how the implementations handle line feeds in single and double quoted strings.

**Comment by dqsully** [comment]
Not quite, I think those railroad diagrams are simplified. In the RFC it allows unescaped characters in the ranges `0x20-0x21` `0x23-0x5B`, and `0x5D-0x10FFFF`. `\n` is `0x10`, so it's not currently allowed to be unescaped

**Comment by dqsully** [comment]
Also, we will need to make sure all implementations continue to support `\r\n` as well as just `\n` as line endings for this new feature.

What should the behavior be if a parser comes across the string `\\\r` without a trailing `\n`? Should we assume it's a line ending and just eat it? Or should we throw an error? I think treating it as an invalid escape sequence makes the most sense, since that's where this feature is probably going to be implemented in our parsers and we can just leverage the existing escape code/framework. e.g. add `\\\r\n` and `\\\n` as valid escape sequences for an empty string or something.

**Comment by trobro** [comment]
I agree, `\r\n` should be supported and removed if escaped, but an escaped `\r` without `\n` should be an error.

**Comment by trobro** [comment]
@AlpyneDreams I had a closer look at the JSON5 specification: it allows `\x`, `\v` and `\0` in strings. And for any unknown escape sequence JSON5 will just remove the backslash (`\A` becomes `A`) Hjson decoders will throw an error when finding unknown escape sequences in single or double quoted strings. Also, U+2028 and U+2029 are valid line terminators in JSON5 but not in Hjson. Similarily JSON5 accepts several unicode whitespace code points that Hjson treats as unquoted strings instead of whitespace. So we will not achieve full JSON5 compatibility.

**Comment by AlpyneDreams** [comment]
> @AlpyneDreams I had a closer look at the JSON5 specification: it allows `\x`, `\v` and `\0` in strings. And for any unknown escape sequence JSON5 will just remove the backslash (`\A` becomes `A`) Hjson decoders will throw an error when finding unknown escape sequences in single or double quoted strings. Also, U+2028 and U+2029 are valid line terminators in JSON5 but not in Hjson. Similarily JSON5 accepts several unicode whitespace code points that Hjson treats as unquoted strings instead of whitespace. So we will not achieve full JSON5 compatibility.

Ah, my mistake, I did not looking too deeply into the minutia. However, I think my point still stands because all of those things feel like much narrower edge cases than multiline continuation (some could possibly be implemented with parse flags too), while multiline continuation appears to me to be the most prominent syntactic difference between valid JSON5 and valid Hjson.

**Comment by trobro** [comment]
I agree, most people won't use those special chars. We just can't officially claim compatibility with JSON5. The differences would require a lot of extra code in Hjson implementations, so I don't think we will do anything about that. But line continuation was easy to fix.

**Comment by isinyaaa** [comment]
Hi @trobro, I know it has long been since we last touched on this subject, but only recently I was reminded about this, so I apologize for the delay!

**Comment by trobro** [comment]
Hi @isinyaaa, thanks for the contribution! Unfortunately noone else has shown interest in contributing, and the Hjson organization is lacking maintainers for most of its repositories. I'm just admin for cpp. So unfortunately I think we cannot introduce general syntax changes in Hjson anymore.

**Comment by isinyaaa** [comment]
> Hi @isinyaaa, thanks for the contribution! Unfortunately noone else has shown interest in contributing, and the Hjson organization is lacking maintainers for most of its repositories.

That's really unfortunate, as it's a very useful tool. I'd like to help out some more if possible, though I particularly don't use it anymore.

> I'm just admin for cpp. So unfortunately I think we cannot introduce general syntax changes in Hjson anymore.

Makes sense to me, I guess we can just leave those PR's open until someone else has found them useful in some way.

**Comment by trobro** [comment]
Nice! I also think Hjson is useful, so I try to keep it alive. @laktak can you invite @isinyaaa to become a member of the Hjson organization?

I agree, keeping the PR's and this issue open makes them visible for others.

**Comment by isinyaaa** [comment]
> Nice! I also think Hjson is useful, so I try to keep it alive. @laktak can you invite @isinyaaa to become a member of the Hjson organization?

I appreciate your appreciation :)

**Comment by laktak** [comment]
Hi @isinyaaa! You should see an invite in your inbox.

**Comment by tbrowder** [comment]
I am not the author of Raku module "JSON::Hjson" (from which I learned  about Hjson, and to use and like it), but I would be happy to try a PR to add links and more information about it and Raku to your website.

**Comment by Jipok** [comment]
Why not simply allow multiline strings in regular quotes? Like this:
```
key: "this is a string
that continues here
and ends here"
```
It would be more natural and readable than using backslashes or triple quotes. Just let the string continue until the closing quote is found, including newlines. This should be backwards compatible with existing Hjson files.

**Comment by trobro** [comment]
> Why not simply allow multiline strings in regular quotes?

In triple quoted strings the newlines become part of the string. When using backslashes for multiline continuation the newlines would not become part of the string. Allowing multiline strings using regular quotes could cause confusion about whether the newlines will be included in the string or not.

**Comment by Jipok** [comment]
> In triple quoted strings the newlines become part of the string.

Well, let's make the same thing true for other types of strings. No backward compatibility will be broken, because in json you can't wrap multiline strings in quotes at all. I don't understand why it was necessary to drag backslashes, because it's a crutch from old languages. Take a look at modern languages ​​like kdl, go, rust, typescript and others. They have exactly this behavior (although not for all types of strings).

**Comment by trobro** [comment]
Triple quotes and backslash continuation solves two different problems. Backslash continuation would allow Hjson files to comply with line length limits (for example 80 chars) and still contain strings longer than 80 chars without line breaks.
